ignite/metrics/nlp/bleu.py

```diff
@@ -122,14 +126,105 @@ def __init__(
         if ngram <= 0:
             raise ValueError(f"ngram order must be greater than zero (got: {ngram})")
         self.ngrams_order = ngram
         self.weights = [1 / self.ngrams_order] * self.ngrams_order
         self.smoother = _Smoother(method=smooth)
+
+        if average not in ["macro", "micro"]:
+            raise ValueError(f'Average must be either "macro" or "micro" (got: {average})')
+        self.average = average
+        self.p_numerators: Counter = Counter()
+        self.p_denominators: Counter = Counter()
+        self.hyp_length_sum = 0
+        self.ref_length_sum = 0
+
         super(Bleu, self).__init__(output_transform=output_transform, device=device)
 
+    def _n_gram_counter(
+        self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],
+    ) -> None:
+        if len(references) != len(candidates):
+            raise ValueError(
+                f"nb of candidates should be equal to nb of reference lists ({len(candidates)} != "
+                f"{len(references)})"
+            )
+
+        # Iterate through each hypothesis and their corresponding references.
+        for refs, hyp in zip(references, candidates):
+            # For each order of ngram, calculate the numerator and
+            # denominator for the corpus-level modified precision.
+            for i in range(1, self.ngrams_order + 1):
+                numerator, denominator = modified_precision(refs, hyp, i)
+                self.p_numerators[i] += numerator
+                self.p_denominators[i] += denominator
+
+            # Calculate the hypothesis lengths
+            self.hyp_length_sum += len(hyp)
+
+            # Calculate the closest reference lengths.
+            self.ref_length_sum += _closest_ref_length(refs, len(hyp))
+
+    def _brevity_penalty_smoothing_macro(self) -> float:
+
+        # Returns 0 if there's no matching n-grams
+        # We only need to check for p_numerators[1] == 0, since if there's
+        # no unigrams, there won't be any higher order ngrams.
+        if self.p_numerators[1] == 0:
+            return 0
+
+        # If no smoother, returns 0 if there's at least one a not matching n-grams
+        if self.smoother.smooth == "no_smooth" and min(self.p_numerators.values()) == 0:
+            return 0
+
+        hyp_len = self.hyp_length_sum
+        ref_len = self.ref_length_sum
+
+        # Calculate corpus-level brevity penalty.
+        if hyp_len < ref_len:
+            bp = math.exp(1 - ref_len / hyp_len) if hyp_len > 0 else 0.0
+        else:
+            bp = 1.0
+        # Smoothing
+        p_n = self.smoother(self.p_numerators, self.p_denominators)
+
+        # Compute the geometric mean
+        s = [w_i * math.log(p_i) for w_i, p_i in zip(self.weights, p_n)]
+        gm = bp * math.exp(math.fsum(s))
+        return gm
+
+    @sync_all_reduce("hyp_lengths", "ref_lengths", "p_numerators", "p_denominators")
+    def _brevity_penalty_smoothing_micro(self) -> float:
```

Comment:
I prefer to have a `compute_micro` and `compute_macro` selected in the `__init__` based on the `average` value.

Reply:
The part of having it selected in the `__init__` throws a mypy error because of [this](https://github.com/python/mypy/issues/708). Instead I have added the if statement in the common `compute` function as you mentioned below.